### ISSUE TITLE
System.Text.Json.Serialization: Converters are ignored for types derived from supported base types

### ISSUE THREAD

**Issue by phizch** [dotnet/corefx#40405]
cc: @layomia,@ahsonkhan, @scalablecory, @steveharter

### Related:
- Add support for types derived from supported BCL collections #39001
- Json serializer support for IEnumerable deserialization extensibilty #36640
- Add support for generic interface-based collections in JsonSerializer #36756
- Json serializer support for collections #36643

### Description

Pull #39001 implemented by @layomia added support for serialization of types that implements natively supported collections without needing to create converters for them.

Unfortunately the implementation treats all derived types as the base type, so it's not possible to create specialized collections with converters. E.g. `class MyCollection : List<int>` gets treated as `List<int>`.

Specialized collections is necessary if the json has multiple ways of declaring a list, e.g.
``` json
{
  "SpecialList" : 
  {
    "Count": 2,
    "Values" : [ 1, 2 ]
  },
  "NormalList" : [ 1, 2 ] 
}
``` 

``` CSharp
public class A
{
    [JsonConverter(typeof(MyCollectionConverter))] // this is ignored and converter for List<int> is used instead
    public MyCollection SpecialList { get; set; }
    public List<int> NormalList { get; set }
}

[JsonConverter(typeof(MyCollectionConverter))] // this is ignored and converter for List<int> is used instead
public class MyCollection : List<int> { }

public class MyCollectionConverter : JsonConverter<MyCollection>
{
    ....
}
public class MyCollectionFactory : JsonConverterFactory
{
	public override bool CanConvert(Type typeToConvert)
    {
        // this only sees List<int>
    }
}
public static void RunTest()
{
    JsonSerializerOptions options = new JsonSerializerOptions
    {
        Converters = 
        { 
            new MyCollectionFactory(), 
            new  MyCollectionConverter() // this is ignored and converter for List<int> is used 
        }
    };

    _ = JsonSerializer.Deserialize<A>("{}", options);
}
```

Expected priorities for getting the converter:
```
// Priority 1: attempt to get converter from JsonConverterAttribute on property.
// Priority 2: Attempt to get custom converter added at runtime.
// Priority 3: Attempt to get converter from [JsonConverter] on the type being converted.
// Priority 4: Attempt to get built-in converter.
```
Current behavior
```
// Priority 0: if type is derived from supported type, use converter for base type
// Priority 1: attempt to get converter from JsonConverterAttribute on property.
// Priority 2: Attempt to get custom converter added at runtime.
// Priority 3: Attempt to get converter from [JsonConverter] on the type being converted.
// Priority 4: Attempt to get built-in converter.
```

From `JsonClassInfo.AddProperty` [source]From `JsonClassInfo.AddProperty` [source](https://github.com/dotnet/corefx/blob/3d72b56b8e7d1e4e44262be6c45c2aac580bbbc4/src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.AddProperty.cs#L30-L41):
``` CSharp
    // Get implemented type, if applicable.
    // Will return the propertyType itself if it's a non-enumerable, string, or natively supported collection.
    Type implementedType = GetImplementedCollectionType(propertyType);
    if (implementedType != propertyType)
    {
        jsonInfo = CreateProperty(implementedType, implementedType, implementedType, null, typeof(object), options);
    }
    else
    {
        jsonInfo = CreateProperty(propertyType, propertyType, propertyType, propertyInfo, classType, options);
    }
```
In my example `GetImplementedCollectionType(typeof(MyCollection))` will return `List<int>`.

I'm currently looking into it to see if I can find a fix.

--
Just found another bug: Since `propertyInfo` is ignored any `JsonIgnoreAttribute` will also be ignored. 
e.g.
```
public class C
{
    public int Version { get; set; }
    
    [JsonIgnore] // this will be ignored
    public MyCollection { get; set; }
}
```

**Comment by layomia** [comment]
@phizch, thanks for the detailed description - I was able to repro this. This is something we should fix for 3.0, and I've opened https://github.com/dotnet/corefx/pull/40411 to address it.

> Just found another bug: Since propertyInfo is ignored any JsonIgnoreAttribute will also be ignored.

This has been fixed in https://github.com/dotnet/corefx/pull/40401.

**Comment by ahsonkhan** [comment]
Re-opening until its fixed for 3.0.

**Comment by layomia** [comment]
Closed in 3.0 by https://github.com/dotnet/corefx/pull/40432.

**Comment by Rauce** [comment]
This bug is marked as fixed, although I am seeing a very similar bug in which my `JsonConverter` is ignored if a property type is `Dictionary<string, string>`.
Here's my original Stack Overflow post, from which I was pointed here: https://stackoverflow.com/questions/58217491/deserialising-json-using-jsonserializer-deserializeasync-is-not-using-my-jsoncon/58217930

**Comment by ahsonkhan** [comment]
@Rauce, can you please file a separate issue for this so we can track and fix the issue you are observing Thanks!

**Comment by ahsonkhan** [comment]
@layomia - can you take a look at this:

> As an aside, it seems like at least one of my cases in this issue still doesn't work;
> 
> ```cs
> public class A
> {
>     [JsonConverter(typeof(MyCollectionConverter))] // no worky
>     public MyCollection SpecialList { get; set; }
>     public List<int> NormalList { get; set; }
> }
> ```
> 
> It's not critical though, since the preferred way is to set the attribute on the `MyCollection` class.

**Comment by steveharter** [comment]
> @layomia - can you take a look at this:
As an aside, it seems like at least one of my cases in this issue still doesn't work;

The JsonConverter attribute not working when applied to a collection is a [bug](https://github.com/dotnet/corefx/issues/41425) that is expected to be fixed for 3.1
